src/jit/MemoryInl.h

```diff
@@ -1011,3 +1011,326 @@ static void emitStore(sljit_compiler* compiler, Instruction* instr)
+    case ByteCode::I64AtomicRmw8AndUOpcode:
+    case ByteCode::I64AtomicRmw16AndUOpcode:
+    case ByteCode::I64AtomicRmw32AndUOpcode: {
+        operation = SLJIT_AND;
+        break;
+    }
+    case ByteCode::I32AtomicRmwOrOpcode:
+    case ByteCode::I32AtomicRmw8OrUOpcode:
+    case ByteCode::I32AtomicRmw16OrUOpcode:
+    case ByteCode::I64AtomicRmwOrOpcode:
+    case ByteCode::I64AtomicRmw8OrUOpcode:
+    case ByteCode::I64AtomicRmw16OrUOpcode:
+    case ByteCode::I64AtomicRmw32OrUOpcode: {
+        operation = SLJIT_OR;
+        break;
+    }
+    case ByteCode::I32AtomicRmwXorOpcode:
+    case ByteCode::I32AtomicRmw8XorUOpcode:
+    case ByteCode::I32AtomicRmw16XorUOpcode:
+    case ByteCode::I64AtomicRmwXorOpcode:
+    case ByteCode::I64AtomicRmw8XorUOpcode:
+    case ByteCode::I64AtomicRmw16XorUOpcode:
+    case ByteCode::I64AtomicRmw32XorUOpcode: {
+        operation = SLJIT_XOR;
+        break;
+    }
+    case ByteCode::I32AtomicRmwXchgOpcode:
+    case ByteCode::I32AtomicRmw8XchgUOpcode:
+    case ByteCode::I32AtomicRmw16XchgUOpcode:
+    case ByteCode::I64AtomicRmwXchgOpcode:
+    case ByteCode::I64AtomicRmw8XchgUOpcode:
+    case ByteCode::I64AtomicRmw16XchgUOpcode:
+    case ByteCode::I64AtomicRmw32XchgUOpcode: {
+        operation = OP_XCHG;
+        break;
+    }
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+        break;
+    }
+
+    AtomicRmw* rmwOperation = reinterpret_cast<AtomicRmw*>(instr->byteCode());
+    offset = rmwOperation->offset();
+
+    Operand* operands = instr->operands();
+    MemAddress addr(options, instr->requiredReg(0), instr->requiredReg(1), 0);
+
+    addr.check(compiler, operands, offset, size);
+
+    JITArg dst;
+    sljit_s32 tmpReg;
+    sljit_s32 srcReg;
+
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+    if ((operationSize & SLJIT_32) || operationSize == SLJIT_MOV32) {
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+        JITArg tmp(operands + 0);
+        JITArg src(operands + 1);
+        dst = JITArg(operands + 2);
+        tmpReg = GET_SOURCE_REG(tmp.arg, instr->requiredReg(0));
+        srcReg = GET_SOURCE_REG(src.arg, instr->requiredReg(1));
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+    } else {
+        JITArgPair tmpPair(operands + 0);
+        JITArgPair srcPair(operands + 1);
+        JITArgPair dstPair(operands + 2);
+        tmpReg = GET_TARGET_REG(tmpPair.arg1, instr->requiredReg(0));
+        srcReg = GET_TARGET_REG(srcPair.arg1, instr->requiredReg(1));
+
+        dst.arg = dstPair.arg1;
+        dst.argw = dstPair.arg1w;
+        sljit_emit_op1(compiler, SLJIT_MOV, dstPair.arg2, dstPair.arg2w, SLJIT_IMM, 0);
+    }
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+
+    struct sljit_label* store_failure = sljit_emit_label(compiler);
+    sljit_emit_atomic_load(compiler, operationSize, tmpReg, GET_SOURCE_REG(addr.memArg.arg, instr->requiredReg(0)));
+    sljit_emit_op1(compiler, SLJIT_MOV, dst.arg, dst.argw, tmpReg, 0);
```

Comment:
This is also something RISCV don't like. Maybe we should add a feature to keep tmpReg on success if this is not always the case (compare and exchange often updates this on fail, but not on success).

Reply:
I mean the RISCV doc says:
An LR/SC sequence begins with an LR instruction and ends with an SC instruction. The dynamic
code executed between the LR and SC instructions can only contain instructions from the base ''I''
instruction set, excluding loads, stores, backward jumps, taken backward branches, JALR, FENCE,
and SYSTEM instructions. If the ''C'' extension is supported, then compressed forms of the
aforementioned ''I'' instructions are also permitted.

This operation might be a memory store in this case.